### PR TITLE
Introduce raw buffer load intrinsic and integrate to the lowering process.

### PR DESCRIPTION
**Pull request by whchung**
Up until now we are using buffer load intrinsic when it comes to vector loads. In PR #248 we started to use raw buffer store intrinsic provided by LLVM AMDGPU, but NOT raw buffer load intrinsic.

@asleepzzz identified while working on https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/138 that it's highly likely we need to switch to raw buffer load intrinsic.

Therefore, in this PR, introduce raw buffer load intrinsic and integrate to the lowering process.

- Fix undesirable `i32` type specification in the syntax of `gpu.buffer_store` and `gpu.raw_buffer_store` ops for shift operand.
- Fix `rawbuf_store.mlir` to REALLY test `gpu.raw_buffer_store` and its lowering to `rocdl.raw.buffer.store`.
- Introduce `gpu.raw_buffer_load`.
- Introduce conversion from `gpu.raw_buffer_load` to `rocdl.raw_buffer_load`.
- Implement unit test to test the conversion process.
- Integrate into `emitLoadInstruction`.

### CODE REVIEW

**Comment by whchung** [comment]
jenkins: retest this please.

**Comment by sjw36** [comment]
Should we just have one op (buffer_load) with an optional attribute 'is_raw'. Looks like the parameters are almost identical (except shift). Same for buffer_store.

**Comment by whchung** [comment]
> Should we just have one op (buffer_load) with an optional attribute 'is_raw'. Looks like the parameters are almost identical (except shift). Same for buffer_store.

The idea is to deprecate `gpu.buffer_load` and `gpu.buffer_store` and eliminate them once `gpu.raw_buffer_load` and `gpu.raw_buffer_store` are fully incorporated in the lowering process.

**Review by sjw36 — APPROVED** [review]
Looks good.

**Comment by whchung** [comment]
@yiqian1 how do set up a CI job to explicitly test this PR with all nightly CI tests? This PR affects every kernel generated by the project so I want to be more cautious before we merge it.

**Comment by yiqian1** [comment]
@whchung Simply copy the existing Jenkins job and change it to use your development branch. I've started two CI jobs:
http://ml-ci.amd.com:21096/job/MLIR/job/test-raw-buffer-load/
http://rocmhead.amd.com:8080/job/test-raw-buffer-load/

**Comment by whchung** [comment]
@asroy now that `emitLoadLogic` and `emitStoreLogic` are largely symmetric, which suggest we are on the right track, but I would like to check the results on nightly CI , as well as getting another review from @asleepzzz before taking this PR.

**Comment by whchung** [comment]
Applying nightly CI for this PR discovered failures:
http://ml-ci.amd.com:21096/job/MLIR/job/test-raw-buffer-load/2/console

From the look of it they are all localized within padding kernel test cases. Need further investigation. Will hold off merging this PR at this moment.

**Comment by whchung** [comment]
A minimal good case:
```
./bin/mlir-miopen-driver -ph -pr -batchsize=8 -in_channels=4 -out_channels=32 -in_h=4 -in_w=4 -fil_h=1 -fil_w=1 --padding_h=1 --padding_w=1 -p=false -c | ./bin/mlir-rocm-runner --shared-libs=./lib/librocm-runtime-wrappers.so,./lib/libmlir_runner_utils.so --entry-point-result=void
```
and a minimal bad case:
```
./bin/mlir-miopen-driver -ph -pr -batchsize=8 -in_channels=3 -out_channels=32 -in_h=4 -in_w=4 -fil_h=1 -fil_w=1 --padding_h=1 --padding_w=1 -p=false -c | ./bin/mlir-rocm-runner --shared-libs=./lib/librocm-runtime-wrappers.so,./lib/libmlir_runner_utils.so --entry-point-result=void
```

`-in_channels` is good at `4`, bad at `3`.

**Comment by whchung** [comment]
Close this PR for now. Reopen / revise it after migrating to OOT.

**Pull request closed without merging**
